lib/sqlalchemy/dialects/mysql/mariadbconnector.py

```diff
@@ -0,0 +1,182 @@
+# <see AUTHORS file>
+#
+# This module is part of SQLAlchemy and is released under
+# the MIT License: http://www.opensource.org/licenses/mit-license.php
+
+"""
+
+.. dialect:: mysql+mariadb
+    :name: MariaDB Connector/Python
+    :dbapi: mariadb
+    :connectstring: mysql+mariadb://<user>:<password>@<host>[:<port>]/<dbname>
+    :url: https://pypi.org/project/mariadb/
+
+Driver Status
+-------------
+
+MariaDB Connector/Python enables python programs to access MariaDB and MySQL
+databases, using an API which is compliant with the Python DB API 2.0 (PEP-249).
+It is written in C and uses MariaDB Connector/C client library for client server
+communication. 
+
+Status: stable
+
+.. mariadb: https://github.com/mariadb-corporation/mariadb-connector-python
+
+"""
+import os
+import re
+
+from .base import MySQLCompiler
+from .base import MySQLDialect
+from .base import MySQLExecutionContext
+from .base import MySQLIdentifierPreparer
+from .base import TEXT
+from ... import sql
+from ... import util
+from distutils.version import StrictVersion
+
+import mariadb.constants.CLIENT as CLIENT
+
+mariadb_cpy_minimum_version="1.0.1"
+
+class MySQLExecutionContext_mariadbconnector(MySQLExecutionContext):
+    pass
+
+class MySQLCompiler_mariadbconnector(MySQLCompiler):
+    pass
+
+class MySQLIdentifierPreparer_mariadbconnector(MySQLIdentifierPreparer):
+    pass
+
+# MariaDB binary protocol doesn't support XA yet, so we need
+# to rewrite the statement. See https://jira.mariadb.org/browse/MDEV-16708
+
+def check_unsupported_xa(statement, parameter):
+    if parameter is None or parameter.__len__() != 1:
+        return None
+    sql = re.sub(re.compile("/\\*.*?\\*/",re.DOTALL ) ,"" ,statement)
+    words= sql.split(None, 1)
+    if words[0].lower() == "xa":
+        if sql.find(" ?"):
+            replace= "'%s'" % parameter[0]
+            return sql.replace("?", replace)
+
+class MySQLDialect_mariadbconnector(MySQLDialect):
+    is_mariadb = True
+    driver = "mariadbconnector"
+    supports_unicode_statements = True
+    encoding = "utf8mb4"
+    convert_unicode = True
+    supports_sane_rowcount = True
+    supports_sane_multi_rowcount = True
+    supports_native_decimal = True
+    default_paramstyle = "qmark"
+    execution_ctx_cls = MySQLExecutionContext_mariadbconnector
+    statement_compiler = MySQLCompiler_mariadbconnector
+    preparer = MySQLIdentifierPreparer_mariadbconnector
+
+    def __init__(self, server_side_cursors=False, **kwargs):
+        super(MySQLDialect_mariadbconnector, self).__init__(**kwargs)
+        self.server_side_cursors = True
+        self.paramstyle= "qmark"
+        if self.dbapi is not None:
+            assert StrictVersion(self.dbapi.__version__) >= StrictVersion(mariadb_cpy_minimum_version),\
+                "The installed version (%s) of MariaDB Connector/Pyton is not supported."\
+                "Please install MariaDB Connector/Python %s or newer" % (self.dbapi.__version__, mariadb_cpy_minimum_version)
+
```

Comment:
**mike bayer** (zzzeek) wrote:

implement _dbapi_version:

    @util.memoized_property
    def _dbapi_version(self):
        if self.dbapi and hasattr(self.dbapi, "__version__"):
            return tuple(
                [
                    int(x)
                    for x in re.findall(
                        r"(\d+)(?:[-\.]?|$)", self.dbapi.__version__
                    )
                ]
            )
        else:
            return (99, 99, 99)

then use:

   if self._dbapi_version < mariadb_cpy_minimum_version:
        raise NotImplementedError("The minimum required version for MariaDB Connector/Python is 1.0.1")

I agree that we should replace all occurrences of the above reg with something nicer someday but I'd prefer to do it all at once.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2166

Reply:
**Gord Thompson** (gordthompson) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2166